### ISSUE TITLE
c++ reverse_iterator: std::find iterates using ptrtoint, instead of ptrs

### ISSUE THREAD

**Issue by adriaanjacobs** [llvm/llvm-project#65743]
Hi

TLDR: SCEV's `getPointerBase` cannot deal with iterations of integer type (for good reason), but `std::find` over an `std::vector` with reverse iterators seems to use an integer (representing a pointer) to iterate over the vector elements at `O2`, which makes it much harder to look through the otherwise simple recurrence in the std::find loop.

When experimenting with some loop optimizations, I saw that std::find in the below snippet would iterate using `ptrtoint`-ed iterators at `-O2`, instead of `ptr` like I expected, and like the forward iterator does. I ran into a problem when trying to use the SCEV interface on this "integer", e.g., `SCEV.getPointerBase` bails out when the type of its argument is not a pointer.

Reproducer on trunk:
```C++
#include <vector>
#include <cstdint>
#include <algorithm>

std::vector<uint64_t> vec;

bool func(uint64_t num) {        
    auto first = vec.rbegin();
    auto last = vec.rend();  
    auto res = std::find(++first, last, num);

    return res != last;        
}
```

As you can see [here in godbolt](https://godbolt.org/z/xoa3Mv9Pz), the iterating "pointer" of the find loop is initialized as follows (at `-O2`):
```llvm
define dso_local noundef zeroext i1 @func(unsigned long)(i64 noundef %num) local_unnamed_addr {
entry:
  %0 = load ptr, ptr getelementptr inbounds ({ { ptr, ptr, ptr } }, ptr @vec, i64 0, i32 0, i32 1), align 8
  %1 = load ptr, ptr @vec, align 8
  %incdec.ptr.i.i = getelementptr inbounds i64, ptr %0, i64 -1
  %2 = ptrtoint ptr %incdec.ptr.i.i to i64
  %3 = ptrtoint ptr %1 to i64
  %sub.ptr.sub.i.i.i.i.i = sub i64 %2, %3
  %cmp79.i.i.i = icmp sgt i64 %sub.ptr.sub.i.i.i.i.i, 31
  br i1 %cmp79.i.i.i, label %for.body.lr.ph.i.i.i, label %for.end.i.i.i

for.body.lr.ph.i.i.i:
  %shr.i.i.i = lshr i64 %sub.ptr.sub.i.i.i.i.i, 5
  br label %for.body.i.i.i

for.body.i.i.i:
  %4 = phi ptr [ %incdec.ptr.i.i, %for.body.lr.ph.i.i.i ], [ %incdec.ptr.i.i.i44.i.i.i, %if.end16.i.i.i ]
  %5 = phi i64 [ %2, %for.body.lr.ph.i.i.i ], [ %11, %if.end16.i.i.i ]
  %__trip_count.080.i.i.i = phi i64 [ %shr.i.i.i, %for.body.lr.ph.i.i.i ], [ %dec.i.i.i, %if.end16.i.i.i ]
  %6 = inttoptr i64 %5 to ptr
  %incdec.ptr.i.i.i.i.i.i = getelementptr inbounds i64, ptr %6, i64 -1
  %7 = load i64, ptr %incdec.ptr.i.i.i.i.i.i, align 8
  %cmp.i.i.i.i = icmp eq i64 %7, %num
; ....
```
The `++first` expression is evaluated into `%incdec.ptr.i.i`, and then `ptrtoint`-ed in `%2` to compute the trip count. This also happens at `-O1`. However, the phinode `%5`, which is semantically the iterating pointer, is of type `i64`, and gets `inttoptr`-converted in the body of the loop to access the vector elements. This seems to differ from the behavior at `-O1`, which keeps the phinode as a pointer type (initialized with `%incdec.ptr.i.i`), and does not rely on `inttoptr` inside the loop. 

The SCEV expression for the `%5` phinode is `{(-8 + (ptrtoint i64* %incdec.ptr.i.i to i64)),+,-32}<nw><%for.body.i.i.i>`, which is an AddRec on which `ScalarEvolution::getPointerBase` immediately bails:
```C++
/// Transitively follow the chain of pointer-type operands until reaching a
/// SCEV that does not have a single pointer operand. This returns a
/// SCEVUnknown pointer for well-formed pointer-type expressions, but corner
/// cases do exist.
const SCEV *ScalarEvolution::getPointerBase(const SCEV *V) {
  // A pointer operand may evaluate to a nonpointer expression, such as null.
  if (!V->getType()->isPointerTy())
    return V;
// ...
}
```

As the comment explains, this might be a corner case. I looked for another interface to look through this construct, but the core issue seems to be that the base of the AddRec is an Add expression (`(-8 + (ptrtoint i64* %incdec.ptr.i.i to i64))`), for which SCEV (or any analysis for that matter) needs to know which side is the pointer and which side is the integer (offset). In this case, I could hardcode a fix for it because the offset is a constant, but more generally, why does the iteration happen using an integer in the first place? 

FYI: as visible in the godbolt, a forward iteration using std::find does not suffer from the same issue. This made me think that std::find using reverse_iterators was simply coded this way, but I could not immediately find source-level `ptrtoint` casts, and the fact that `O1` does not do this (even for reverse_iterators) also throws me off.

**Comment by efriedma-quic** [comment]
The point of getPointerBase is to get the base pointer of an expression.  If the expression doesn't have pointer type, that doesn't make any sense; there is no pointer, so there's no base pointer.  At best, you'd have a heuristic to try to find the most complex operand of an expression.

Really, your problem here isn't the ptrtoint; SCEV computations involving it generally works out fine (including loop trip counts).  The problem is the inttoptr; LLVM optimizations are extremely conservative with inttoptr ops because the semantics are weird.

It looks like the inttoptr is getting generated by SROA, after instcombine lowers a memcpy to an i64 load/store pair.  That instcombine optimization is known to be somewhat problematic, but we don't have a good substitute.

**Comment by adriaanjacobs** [comment]
Looking through the optimization passes applied to the snippet, I'm not immediately able to spot the memcpy lowering you are talking about. I am able to find a number of ptrtoint/inttoptr conversions happening during iterator copies and equality checks, however. 

I also see that only part of the unrolled equality tests searching for num are done using the integer pointer. Both phi nodes in the IR snippet represent the same pointer, but only one of them actually has the pointer type. 

Could you point me towards the problematic instcombine optimization you are talking about?

**Comment by efriedma-quic** [comment]
https://github.com/llvm/llvm-project/blob/93bd428742f985a4b909dd5efee462ea520c96c0/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp#L153
